### PR TITLE
Helpers arr

### PR DESCRIPTION
**Pull request by dragoonchang**
https://github.com/goravel/goravel/issues/82

### CODE REVIEW

File: support/arr/arr.go [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+			builder.WriteString(finalSeparator[0])
+		} else if i < len(arr)-1 {
+			builder.WriteString(delimiter)
+		}
+	}
+	return builder.String()
+}
+
+// KeyBy Key an associative array by a field or using a callback.
+func KeyBy[T any, K comparable](array []T, key K) ([]T, error) {
```
Comment (hwbrzzl) [inline support/arr/arr.go:341]: S: We can remove it.

File: support/arr/arr.go [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+	return nil, ErrNoImplementation
+}
+
+// PrependKeysWith Prepend the key names of an associative array.
+func PrependKeysWith[T any](array []T, prependWith string) ([]T, error) {
```
Comment (hwbrzzl) [inline support/arr/arr.go:346]: S: We can remove it, implement it in map.

**Review by hwbrzzl — COMMENTED** [review]
A great PR 👍 
Haven't finished the review yet, I'll continue when I'm free.

File: support/arr/arr.go [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+	return nil, ErrNoImplementation
+}
+
+// Only returns a subset of the items from the given map with specified keys.
+func Only[T any](arr []T, keys any) []T {
```
Comment (hwbrzzl) [inline support/arr/arr.go:351]:
```suggestion
func Only[T any, K int|[]int](arr []T, keys K) []T {
```

File: support/arr/arr.go [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+				return res
+			}
+			res = append(res, arr[vv])
+		}
+	}
+	return res
+}
+
+// Pluck an array of values from an array.
+func Pluck[T any](array []T, value T, key ...int) ([]T, error) {
```
Comment (hwbrzzl) [inline support/arr/arr.go:372]: S: We can remove it.

File: support/arr/arr.go [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+	return nil, ErrNoImplementation
+}
+
+// ExplodePluckParameters Explode the "value" and "key" arguments passed to "pluck".
+func ExplodePluckParameters[T any](array []T, key ...int) ([]T, error) {
```
Comment (hwbrzzl) [inline support/arr/arr.go:377]: S: We can remove it.

File: support/arr/arr.go [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+	return res
+}
+
+// Prepend the given value to the beginning of an array or associative array.
+func Prepend[T any](arr []T, value T) []T {
+	return append([]T{value}, arr...)
+}
+
+// Pull Get a value from the array, and remove it.
+func Pull[T any](arr []T, key int, def T) ([]T, T, error) {
```
Comment (hwbrzzl) [inline support/arr/arr.go:396]:
```suggestion
func Pull[T any](arr []T, key int, def ...T) (T, error) {
```

File: support/arr/arr.go [hunk 7]
```diff
@@ -0,0 +1,633 @@
...
+	v, _ := Get(arr, key, def)
+
+	res, err := Forget(arr, key)
+	return res, v, err
+}
+
+// Query Convert the array into a query string.
+func Query[T any](m []T) (*string, error) {
```
Comment (hwbrzzl) [inline support/arr/arr.go:404]: S: We can remove it, implement it in map.

File: support/arr/arr.go [hunk 8]
```diff
@@ -0,0 +1,633 @@
...
+	return nil, ErrNoImplementation
+}
+
+// Random returns one or a specified number of random values from a slice.
+func Random[T any](arr []T, number *int) ([]T, error) {
```
Comment (hwbrzzl) [inline support/arr/arr.go:409]:
```suggestion
func Random[T any](arr []T) (T, error) {}
func Randoms[T any](arr []T, number int) ([]T, error) {}
```

File: support/arr/arr.go [hunk 9]
```diff
@@ -0,0 +1,633 @@
...
+	//rand.Seed(time.Now().UnixNano())
```
Comment (hwbrzzl) [inline support/arr/arr.go:410]: Q: remove it?

File: support/arr/arr.go [hunk 10]
```diff
@@ -0,0 +1,633 @@
...
+	for _, index := range indices {
+		res = append(res, arr[index])
+	}
+
+	return res, nil
+}
+
+// Set an array item to a given value using int key
+// todo: thread safe?
+func Set[T any](arr *[]T, key int, value T) error {
```
Comment (hwbrzzl) [inline support/arr/arr.go:443]: This's ok 👍

File: support/arr/arr.go [hunk 11]
```diff
@@ -0,0 +1,633 @@
...
+		copy(newSlice, *arr)
+		*arr = newSlice
+	}
+
+	(*arr)[key] = value
+	return nil
+}
+
+// Shuffle the given array and return the result.
+func Shuffle[T any](arr []T, seed *int64) []T {
```
Comment (hwbrzzl) [inline support/arr/arr.go:459]: Q: Can we remove the seed parameter?
```suggestion
func Shuffle[T any](arr []T) []T {
```

File: support/arr/arr.go [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+	} else {
+		r = rand.New(rand.NewSource(*seed))
+	}
+	r.Shuffle(len(res), func(i, j int) { res[i], res[j] = res[j], res[i] })
+	return res
+}
+
+// Sort the nested array using the given callback.
+// todo: generic
+func Sort(arr []any, fn func(i, j int) bool) []any {
```
Comment (hwbrzzl) [inline support/arr/arr.go:476]:
```suggestion
func Sort[T any](arr []T, fn func(i, j int) bool) []T {
```

File: support/arr/arr.go [hunk 13]
```diff
@@ -0,0 +1,633 @@
...
+		default:
+		}
+	}
+
+	return arr
+}
+
+// Sort the nested array in descending order using the given callback.
+// todo: generic
+func SortDesc(arr []any, fn func(i, j int) bool) []any {
```
Comment (hwbrzzl) [inline support/arr/arr.go:498]:
```suggestion
func SortDesc[T any](arr []T, fn func(i, j int) bool) []T {
```

File: support/arr/arr.go [hunk 14]
```diff
@@ -0,0 +1,633 @@
...
+	return Sort(arr, fn)
+}
+
+// SortRecursive Recursively sort an array by values.
+// todo: generic
+func SortRecursive(arr []any, descending bool) ([]any, error) {
```
Comment (hwbrzzl) [inline support/arr/arr.go:504]: S: A great method, but if we can't use generic or will more complex, I think we can not implement it first.

File: support/arr/arr.go [hunk 15]
```diff
@@ -0,0 +1,633 @@
...
+		} else {
+			return fn(res[j], res[i])
+		}
+	})
+
+	return res, nil
+}
+
+// ToCssClasses Convert an array of strings to a string of CSS classes.
+func ToCssClasses[T any](arr []T) string {
```
Comment (hwbrzzl) [inline support/arr/arr.go:544]: S: Maybe we should move this method to `map`, like `func ToCssClasses(arr map[string]bool) string {}`

File: support/arr/arr.go [hunk 16]
```diff
@@ -0,0 +1,633 @@
...
+
+	for _, v := range arr {
+		res = append(res, fmt.Sprint(v))
+	}
+
+	return strings.Join(res, " ")
+}
+
+// ToCssStyles Convert an array of strings to a string of CSS styles.
+func ToCssStyles[T any](arr []T) string {
```
Comment (hwbrzzl) [inline support/arr/arr.go:555]: S: Maybe we should move this method to map, like func ToCssStyles(arr map[string]bool) string {}

File: support/arr/arr.go [hunk 17]
```diff
@@ -0,0 +1,633 @@
...
+	notNilFilter := func(item T) bool {
+		rv := reflect.ValueOf(item)
+		return rv.IsValid() && !rv.IsZero()
+	}
+	return Where(arr, notNilFilter)
+}
+
+// Wrap If the given value is not an array and not null, wrap it in one.
+// todo: generic
+func Wrap(value any) []any {
```
Comment (hwbrzzl) [inline support/arr/arr.go:589]: S: Mabe we don't need this method in Golang.

**Review by hwbrzzl — COMMENTED** [review]
Hi @dragoonchang  Thanks for your hard work on the helper method, for more efficiency, I suggest we can make the method structure first(no body). We can implement the logic after we get the consensus. 
Thanks again!

File: support/arr/arr.go [hunk 18]
```diff
@@ -0,0 +1,606 @@
+package arr
+
+import (
+	"fmt"
+	"github.com/goravel/framework/support/time"
+	"math/rand"
+	"reflect"
+	"sort"
+	"strings"
```
[the remark this answers is not in the document]
Comment (dragoonchang) [reply support/arr/arr.go]: done

File: support/arr/arr.go [hunk 19]
```diff
@@ -0,0 +1,633 @@
...
+		return []any{}
+	}
+	res := make([]any, 0)
+	recursiveCollapse(arr, &res)
+	return res
+}
+
+func recursiveCollapse(value any, res *[]any) {
+	switch v := value.(type) {
+	case [][]interface{}:
```
[the remark this answers is not in the document]
Comment (dragoonchang) [reply support/arr/arr.go:46]: done

[the remark this answers is not in the document]
Comment (dragoonchang) [reply support/arr/arr.go:46]: It won't work if interface{} to any, please try it and suggest.

File: support/arr/arr.go [hunk 20]
```diff
@@ -0,0 +1,633 @@
...
+	}
+}
+
+// CrossJoin returns all possible permutations of the given arrays.
+func CrossJoin[T any](arr ...[]T) ([][]T, error) {
+	if len(arr) == 0 {
+		return nil, ErrArrayRequired
+	}
+
+	res := [][]T{{}}
```
[the remark this answers is not in the document]
Comment (dragoonchang) [reply support/arr/arr.go:81]: It won't work if interface{} to any, please try it and suggest.

File: support/arr/arr.go [hunk 21]
```diff
@@ -0,0 +1,633 @@
...
+		if err := Set(&arr, key, value); err != nil {
+			return arr, err
+		}
+	}
+
+	return arr, nil
+}
+
+// Collapse collapses an array of arrays into a single array.
+func Collapse(arr []any) []any {
```
[the remark this answers is not in the document]
Comment (dragoonchang) [reply support/arr/arr.go]: done

File: support/arr/arr.go [hunk 22]
```diff
@@ -0,0 +1,633 @@
...
+	return nil, ErrNoImplementation
+}
+
+// Undot returns an expanded array from flattened dot notation array
+func Undot[T any](m []T) ([]T, error) {
+	return nil, ErrNoImplementation
+}
+
+// Except returns all the given array except for a specified array of keys.
+func Except[T any](arr []T, keys []int) []T {
```
[the remark this answers is not in the document]
Comment (dragoonchang) [reply support/arr/arr.go]: done

File: support/arr/arr.go [hunk 23]
```diff
@@ -0,0 +1,633 @@
...
+		if v < 0 || v >= len(arr) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// HasAny Determine if any of the keys exist in an array using int key
+func HasAny[T any](arr []T, keys any) bool {
```
[the remark this answers is not in the document]
Comment (dragoonchang) [reply support/arr/arr.go]: done

File: support/arr/arr.go [hunk 24]
```diff
@@ -0,0 +1,633 @@
...
+func Join[T any](arr []T, delimiter string, finalSeparator ...string) string {
+	l := len(arr)
+	if l == 0 {
+		return ""
+	}
+	if l == 1 {
+		return fmt.Sprint(arr[0])
+	}
+	if l == 2 {
+		return fmt.Sprintf("%v%s%v", arr[0], finalSeparator[0], arr[1])
```
[the remark this answers is not in the document]
Comment (dragoonchang) [reply support/arr/arr.go:354]: Do you mean I should make it panic or avoid it? 
Below code will prevent panic
	if l == 0 {
		return ""
	}
also below is the corresponding test case:
	arr5 := []string{}
	expectedStr5 := ""
	result5 := Join(arr5, ", ", " and ")
	assert.Equal(t, expectedStr5, result5)

File: support/arr/arr.go (hunk 1 above)
Comment (dragoonchang) [reply support/arr/arr.go]: done

File: support/arr/arr.go (hunk 2 above)
Comment (dragoonchang) [reply support/arr/arr.go]: done

File: support/arr/arr.go (hunk 4 above)
Comment (dragoonchang) [reply support/arr/arr.go]: done

File: support/arr/arr.go (hunk 5 above)
Comment (dragoonchang) [reply support/arr/arr.go]: done

**Comment by dragoonchang** [comment]
> Hi @dragoonchang Thanks for your hard work on the helper method, for more efficiency, I suggest we can make the method structure first(no body). We can implement the logic after we get the consensus. Thanks again!

That sounds more efficient! I will open another file for this purpose.

**Comment by hwbrzzl** [comment]
Thanks for your optimization, can I review this PR again, please?

**Comment by hwbrzzl** [comment]
Hi @dragoonchang  I added a new label: Ready For Review, you can add it when this PR can be reviewed.

**Pull request closed without merging**